File: foodapp/src/main/java/com/kspk/Utils/JwtUtils.java
package com.kspk.Utils;

import io.jsonwebtoken.Claims;
import io.jsonwebtoken.Jwts;
import io.jsonwebtoken.io.Decoders;
import io.jsonwebtoken.security.Keys;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.stereotype.Service;

import javax.crypto.SecretKey;
import java.util.Date;
import java.util.HashMap;
import java.util.Map;
import java.util.function.Function;

@Service
public class JwtUtils {

    // 1. Corrected to use the standard Spring property name
    @Value("${jwt.secret}")
    private String jwtSecret;

    // 2. Added property for token expiration time (default to 24 hours)
    @Value("${jwt.expiration.ms:86400000}")
    private long jwtExpirationMs;

    // Helper to get the SecretKey object from the encoded string
    private SecretKey getSigningKey() {
        byte[] keyBytes = Decoders.BASE64.decode(jwtSecret);
        return Keys.hmacShaKeyFor(keyBytes);
    }

    public String generateToken(UserDetails userDetails){
        // 3. Corrected generic type to Map<String, Object>
        Map<String, Object> claims = new HashMap<>();
        return createToken(claims, userDetails.getUsername());
    }

    private String createToken(Map<String, Object> claims, String subject) {
        SecretKey key = getSigningKey();
        return Jwts.builder()
                .setClaims(claims)
                .setSubject(subject)
                .setIssuedAt(new Date(System.currentTimeMillis()))
                // Use the configured expiration time
                .setExpiration(new Date(System.currentTimeMillis() + jwtExpirationMs))
                .signWith(key)
                .compact();
    }

    public String extractUsername(String token){
        return extractClaim(token, Claims::getSubject);
    }

    public Date extractExpiration(String token){
        return extractClaim(token, Claims::getExpiration);
    }

    public <T> T extractClaim(String token, Function<Claims, T> claimResolver){
        final Claims claims = extractAllClaims(token);
        return claimResolver.apply(claims);
    }

    private Claims extractAllClaims(String token){
        // 4. Use the SecretKey object for parsing
        return Jwts.parserBuilder()
                .setSigningKey(getSigningKey())
                .build()
                .parseClaimsJws(token)
                .getBody();
    }

    private Boolean isTokenExpired(String token){
        return extractExpiration(token).before(new Date());
    }

    public Boolean validateToken(String token, UserDetails userDetails){
        final String username = extractUsername(token);

        return (username.equals(userDetails.getUsername()) && !isTokenExpired(token));
    }
}

File: food-app-backend-js/src/utils/jwt.utils.js
import jwt from 'jsonwebtoken';

const SECRET_KEY = process.env.JWT_SECRET || 'default_secret';
const EXPIRATION_SECONDS = parseInt(process.env.JWT_EXPIRATION_MS || '86400000') / 1000;

class JwtUtils {
  generateToken(user) {
    const payload = {
      sub: user.email,
      role: user.role,
    };
    return jwt.sign(payload, SECRET_KEY, { expiresIn: EXPIRATION_SECONDS });
  }

  extractUsername(token) {
    try {
      const decoded = jwt.decode(token);
      return decoded ? decoded.sub : null;
    } catch (e) {
      return null;
    }
  }

  extractRole(token) {
    try {
      const decoded = jwt.decode(token);
      return decoded ? decoded.role : null;
    } catch (e) {
      return null;
    }
  }

  validateToken(token, user) {
    try {
      const decoded = jwt.verify(token, SECRET_KEY);
      return decoded.sub === user.email;
    } catch (e) {
      return false;
    }
  }
}

export default new JwtUtils();
